### PR TITLE
added recent activity date col for DDAH table

### LINKED ISSUE #549: Recent Activity Date missing from DDAH view
A recent activity column should be added to the DDAH table.

### PR DESCRIPTION
**Pull request by Altair59**
#549 added recent activity date col for DDAH table

### CODE REVIEW

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 1]
```diff
@@ -54,6 +55,33 @@ export function getReadableStatus(ddah: Pick<Ddah, "status">) {
...
+
+function getRecentActivityDate(ddah: Ddah) {
+    const validDates = [
+        ddah.accepted_date,
+        ddah.approved_date,
+        ddah.emailed_date,
+        ddah.revised_date,
+    ]
+        .filter((dateString) => {
+            return isIsoDateString(dateString);
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: Just do `filter(date => date)`. There's no need to validate.

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 2]
```diff
@@ -54,6 +55,33 @@ export function getReadableStatus(ddah: Pick<Ddah, "status">) {
     }
 }
 
+function isIsoDateString(dateString: string | null) {
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: Get rid of this function.

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 3]
```diff
@@ -54,6 +55,33 @@ export function getReadableStatus(ddah: Pick<Ddah, "status">) {
...
+        ddah.revised_date,
+    ]
+        .filter((dateString) => {
+            return isIsoDateString(dateString);
+        })
+        // dateString || 0 is required to avoid
+        // typescript complaining about potentially passing null to Date constructor,
+        // which cannot happen (although it complains) due to previous filtering
+        .map((dateString) => new Date(dateString || 0).getTime());
+    if (validDates.length === 0) return "";
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: Never leave off `{  }` on an if statement.

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 4]
```diff
@@ -54,6 +55,33 @@ export function getReadableStatus(ddah: Pick<Ddah, "status">) {
...
+        ddah.emailed_date,
+        ddah.revised_date,
+    ]
+        .filter((dateString) => {
+            return isIsoDateString(dateString);
+        })
+        // dateString || 0 is required to avoid
+        // typescript complaining about potentially passing null to Date constructor,
+        // which cannot happen (although it complains) due to previous filtering
+        .map((dateString) => new Date(dateString || 0).getTime());
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: What? Dates can already be compared numerically without a call to `getTime`

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 5]
```diff
@@ -54,6 +55,33 @@ export function getReadableStatus(ddah: Pick<Ddah, "status">) {
...
+    if (validDates.length === 0) return "";
+    const recentActivityDate = new Date(Math.max.apply(null, validDates));
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: How about `new Date(Math.max(...[<all the dates>].filter(date => date).map(date => new Date(date))))`

File: frontend/src/views/admin/ddah-table/index.tsx [hunk 6]
```diff
@@ -25,6 +25,7 @@ export interface RowData {
     first_name: string;
     total_hours: number | null;
     status: string | null;
+    recent_activity_date: string;
```
Comment (siefkenj) [inline frontend/src/views/admin/ddah-table/index.tsx]: Should be nullable if there is no recent activity.

File: frontend/src/views/admin/ddah-table/index.tsx (hunk 5 above)
Comment (Altair59) [reply frontend/src/views/admin/ddah-table/index.tsx]: But Typescript will complain that `Argument of type 'Date' is not assignable to parameter of type 'number'` when using `Math.max`, which is why I convert `Date`s to numeric values using `getTime`

Comment (siefkenj) [reply frontend/src/views/admin/ddah-table/index.tsx]: In that case, you can cast them all to numbers (use `+` as a prefix to cast), or you can retype `Max`. E.g.,

```
let x = new Date();
let y = new Date();

let z =(Math.max as (...values: any[]) => number)(x,y);
```

**Pull request merged**